### PR TITLE
Add support Drupal 8

### PR DESCRIPTION
**Pull request by ghost**
Included ajax validation for API credentials
Includes a hook_help() for support/help.
Added more cURL error code checks with messages

### CODE REVIEW

File: smaily_for_drupal/smaily_for_drupal.module [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Implements hook_help().
+ */
+function smaily_for_drupal_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    case 'help.page.smaily_for_drupal':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('The Smaily for Drupal module allows you to integrate the Smaily API for your site. A newsletter subscription block can be configured and placed where possible.<br> More information about Smaily is available at the Smaily <a href=http://help.smaily.com/en/support/home> Knowledgebase.') . '</a></a></p>';
```
Comment (kaittodesk) [inline smaily_for_drupal/smaily_for_drupal.module]: Try to avoid having links inside translation phrases.

This can be written as:
```php
t('... More information about Smaily is available in the Smaily <a href=":knowledgebase_url">Knowledgebase</a>', array(':knowledgebase_url' => 'http://help.smaily.com/en/support/home'));
```

PS. You have 2 link closing tags outside of translation phrase. One should be removed and the other should be in the phrase.

File: smaily_for_drupal/smaily_for_drupal.module [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      $output .= '<h3>' . t('Uses') . '</h3>';
+      $output .= '<dl>';
+      $output .= '<dt>' . t('Configuring Smaily API credentials') . '</dt>';
+      $output .= '<dd>' . t("Fill out and validate the credentials provided to you by Smaily in the administration page.") . '</dd>';
```
Comment (kaittodesk) [inline smaily_for_drupal/smaily_for_drupal.module]: Formatting: be consistent when using double or single quotes.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+use Drupal\Core\Form\ConfigFormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\HtmlCommand;
+use Drupal\Core\Ajax\ReplaceCommand;
+
+/**
+ * Smaily module admin configuraiton form.
+ */
+class SmailyForDrupalAdminForm extends ConfigFormBase {
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: I think this can just be `AdminForm`. The `namespace` value should make sure that this does not conflict with any other module.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEditableConfigNames() {
+    return [
+      'smaily_for_drupal.adminsettings',
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: One could think that `adminsettings` refers to admin form settings, e.g. these settings are for configuring the admin form not `Smaily for Drupal` module. I would go with `smaily_for_drupal.settings`, as these are this module's settings.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+      '#type' => 'fieldset',
+      '#title' => $this->t('API credentials'),
+    ];
+
+    $form['smaily_container']['credentials']['message'] = [
+      '#type' => 'markup',
+      '#markup' => '<div class="smaily_message"></div>',
+    ];
+
+    $form['smaily_container']['credentials']['smaily_domain'] = [
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: I would go with `['smaily_api_credentials']['domain']` (and similar) grouping. It makes more sense.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+      '#type' => 'textfield',
+      '#title' => $this->t('Domain'),
+      '#default_value' => $config->get('smaily_domain', ''),
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: And I would change configuration option key from `smaily_domain` to `smaily_api_credentials.domain` (and apply the same pattern to rest of the configuration options), it is more explicit and avoids possible conflicts in the future. Hopefully the key is not too long.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    ];
+
+    $username = $config->get('smaily_username');
+    $password = $config->get('smaily_password');
+    $domain = $config->get('smaily_domain');
+
+    // Fetch and update autoresponders every time form is built.
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: What happens if `smaily_domain` does not have a value?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    $form['smaily_container']['smaily_button_title'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Subscribe button title'),
+      '#default_value' => $config->get('smaily_button_title', 'Subscribe to newsletter'),
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save'),
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: I think it is better if the button has information about what the user is saving. Please rename it to "Save settings".

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+      '#default_value' => $config->get('smaily_username', ''),
+      '#required' => TRUE,
+    ];
+
+    $form['smaily_container']['credentials']['smaily_password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $config->get('smaily_password', ''),
+      '#required' => TRUE,
+
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Formatting: unnecessary whitespace.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    // Fetch and update autoresponders every time form is built.
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
+    );
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+    curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
+    $autoresponders = json_decode(curl_exec($ch), TRUE);
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: What happens if `curl_exec` returns `FALSE`?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    // Validation handled by #required and in ajax function.
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Is this method required here even if it does nothing?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+   * {@inheritdoc}
+   */
+  public function validateCredentials(array &$form, FormStateInterface $form_state) {
+    $ajax_response = new AjaxResponse();
+
+    $username = trim($form_state->getValue('smaily_username'));
+    $password = trim($form_state->getValue('smaily_password'));
+    $subdomain = trim($form_state->getValue('smaily_domain'));
+    // Drupal's validation is only done with submit.
+    if (empty($username) || empty($password) || empty($subdomain)) {
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Isn't this already done by basic form validation? Or doesn't `ConfigFormBase` handle required form fields with AJAX requests?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+          '.smaily_message',
+          '<div class="messages messages--error">' . 'Please fill out all fields.' . '</div>')
+      );
+      return $ajax_response;
+    }
+
+    // Normalize subdomain.
+    // First, try to parse as full URL.
+    // If that fails, try to parse as subdomain.sendsmaily.net, and, then clean up subdomain and pass as is.
+    if (filter_var($subdomain, FILTER_VALIDATE_URL)) {
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Please extract this functionality into a separate method that takes the unformatted subdomain as its value and returns formatted subdomain.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+      $parts = explode('.', $url['host']);
+      $subdomain = count($parts) >= 3 ? $parts[0] : '';
+    }
+    elseif (preg_match('/^[^\.]+\.sendsmaily\.net$/', $subdomain)) {
+      $parts = explode('.', $subdomain);
+      $subdomain = $parts[0];
+    }
+    $subdomain = preg_replace('/[^a-zA-Z0-9]+/', '', $subdomain);
+
+    $ch = curl_init();
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Please extract automations fetching into separate method.

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+namespace Drupal\smaily_for_drupal\Form;
+
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Smaily newsletter signup form.
+ */
+class SmailyForDrupalForm extends FormBase {
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: I think this class can be renamed to `SubscribeForm` (or similar). Again conflicts shouldn't be an issue due to the namespacing.

**Comment by ghost** [comment]
1. I currently added the Drupal 8 module in folder smaily_for_drupal and added a GitHub readme for Drupal 8. I think we should have the same folder versioning logic like in OpenCart.
2. Do you know if and how I could somehow cut out the cURL portion into a controller or class to not repeat it 3 times in the admin form, or is it OK? Smaily is moving to captcha eventually but still would like to know.

Ready for review.

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+      $query_data['name'] = $name;
+    }
+
+    $autoresponder = $config->get('smaily_autoresponder');
+    if (!empty($autoresponder)) {
+      $query_data['autoresponder'] = $autoresponder;
+    }
+
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL, 'https://' . $domain . '.sendsmaily.net/api/opt-in/');
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: Use API endpoint `/api/autoresponder.php`. Please take a look at the [API documentation](http://help.smaily.com/en/support/solutions/articles/16000021289-autoresponder).

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+      ->save();
+    $ajax_response->addCommand(
+      new HtmlCommand(
+        '.smaily_message',
+        '<div class="messages messages--status">' . 'Credentials valid' . '</div>')
+    );
+    return $ajax_response;
+
+  }
+
```
Comment (kaittodesk) [inline smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Is this whitespace a Drupal coding standard?

File: README.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
+# Smaily Drupal module
```
Comment (kaittodesk) [inline README.md]: Use just "Smaily for Drupal".

File: README.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+Smaily email marketing and automation extension plugin for Drupal.
```
Comment (kaittodesk) [inline README.md]: "extension plugin" mean the same thing, but they do not mean anything in Drupal's context. Drupal uses "module" instead.

File: README.md [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+Includes a simple sign-up form to add new subscribers to Smaily subscribers list.
+## Features
```
Comment (kaittodesk) [inline README.md:6]: Add a newline before headings for better readability.

File: README.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+### Drupal Newsletter Subscription
+- Add new subscribers to Smaily subscribers list.
+- Select autoresponder to send automated emails.
+
+## Requirements
+- Drupal 8+
```
Comment (kaittodesk) [inline README.md]: Do not add plus sign after Drupal 8, because that would mean that this module works with future Drupal 9 and later.

File: README.md [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+## Contribution
+All development for Smaily for Drupal is handled via [GitHub](https://github.com/sendsmaily/smaily-drupal-module).
+
+Opening new issues and submitting pull requests are welcome.
+
+## Installation
+
+- Place smaily_for_drupal into core/modules/ folder.
+- Install module under Extend.
+- Validate and save your credentials.
```
Comment (kaittodesk) [inline README.md]: Please extend this a little - Validate and save your Smaily API credentials.

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 12 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: It does not, no. AJAX requests need special validation

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 7 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: cURL will say: Could not resolve host: .sendsmaily.net

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 11 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Not required by ConfigFormBase, no.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 10 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: I tried `json_decode(FALSE, TRUE)` it returned NULL
`if (!empty($autoresponders))` will be skipped

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 17 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Line 276 yes.

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php (hunk 16 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: What's the difference between this and `/api/opt-in/`?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 7 above)
Comment (kaittodesk) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: It would be better to not make any request if any of the API credentials are missing. It is a great way to increasing efficiency.

**Comment by ghost** [comment]
Question. What should be this module's parent when you are on admin/config?(Development/Media/Etc)
It's currently sitting in Web Services with RSS publishing (Drupal core).

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 10 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: Is it bad not to handle a FALSE message and rely on json_decode returning NULL?

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 7 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: The validation would stop any half-filled api credentials from being saved, even with Ajax.
Is it a just-in-case scenario? I could only see this happening when someone manually edits the mySQL table

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php (hunk 16 above)
Comment (kaittodesk) [reply smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: There is none. But why then configure Smaily API credentials if you are not going to use them? :)

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 17 above)
Comment (kaittodesk) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: OK.

**Comment by kaittodesk** [comment]
> What should be this module's parent when you are on admin/config?(Development/Media/Etc)
> It's currently sitting in Web Services with RSS publishing (Drupal core).

I think it is fine. Our competition uses the same section.

**Comment by kaittodesk** [comment]
> Do you know if and how I could somehow cut out the cURL portion into a controller or class to not repeat it 3 times in the admin form, or is it OK? Smaily is moving to captcha eventually but still would like to know.

If you are feeling that the code becomes heavy (too much code for a single function or a file) or logic starts to repeat, then it is always a good idea to start refactoring it, and in the process you might need to create separate classes/functions/methods or even need to split the code into multiple files.

With the current cURL implementation, you could build a separate model for that - a single purpose class that takes in arguments necessary to make the API request and handles the response. Something similar to the Request class you built for Smaily for Wordpress 2.x plugin.

File: smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php (hunk 7 above)
Comment (kaittodesk) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: OK. Maybe I don't understand how and when `buildForm` is called. Doesn't it get called when the form is rendered in the admin panel? If it does, wouldn't it still call `fetchAutoresponders` even if no settings have been saved before?

Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php]: I added an `!empty($domain)` check. Added it now for all credentials
It would still call `fetchAutoresponders` but the `if` check wouldn't start a cURL session.

File: smaily_for_drupal/smaily_for_drupal.module [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Implements hook_help().
+ */
+function smaily_for_drupal_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    case 'help.page.smaily_for_drupal':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('The Smaily for Drupal module allows you to integrate the Smaily API for your site. A newsletter subscription block can be configured and placed where possible.');
```
Comment (kaittodesk) [inline smaily_for_drupal/smaily_for_drupal.module:18]: You are starting a paragraph here, but I do not see a closing tag anywhere.

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php (hunk 16 above)
Comment (ghost) [reply smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: I reworked it for autoresponder.php. Autoresponder.php asks for some authorization while /opt-in/ doesn't, yes? And I already went through the trouble of getting API credentials from config, so might as well cURL to autoresponder.php
Is the current solution OK or did I misunderstand you?

File: smaily_for_drupal/smaily_for_drupal.module (hunk 23 above)
Comment (ghost) [reply smaily_for_drupal/smaily_for_drupal.module:18]: Good eye. Is there any reason for this kind of formatting Drupal uses, or is it the only way to do it?

File: smaily_for_drupal/src/Form/SmailyForDrupalForm.php (hunk 16 above)
Comment (kaittodesk) [reply smaily_for_drupal/src/Form/SmailyForDrupalForm.php]: Yes, that is mighty fine. Thanks!

`/api/autoresponder.php` uses authentication and it is more meant for background requests.  `/api/opt-in/` is more meant for embedded opt-in forms that directly submit form's contents to Smaily. We are trying to make this separation at the moment with all of our integrations.

File: smaily_for_drupal/smaily_for_drupal.module (hunk 23 above)
Comment (kaittodesk) [reply smaily_for_drupal/smaily_for_drupal.module:18]: I am not really sure. Drupal has a decent (although complicated) templating logic. Maybe `hook_help` is triggered in such an early stage of bootstraping that usage of template is not possible. If you'd like you can take a deep dive into how `hook_help` works, but for now I think it will do.

**Review by kaittodesk — APPROVED** [review]
